universal-application-tool-0.0.1/app/services/program/ProgramService.java

```diff
@@ -1,17 +1,88 @@
 package services.program;
 
 import com.google.common.collect.ImmutableList;
+import java.util.Optional;
 
-/** Operations you can perform on Programs. */
+/** Operations you can perform on {@link ProgramDefinition}s. */
 public interface ProgramService {
 
-    /** List all programs. */
-    public ImmutableList<ProgramDefinition> listProgramDefinitions();
+  /**
+   * List all programs.
+   *
+   * @return a list of {@link ProgramDefinition}s
+   */
+  ImmutableList<ProgramDefinition> listProgramDefinitions();
 
-    /** Get the definition for a given program. */
-    public ProgramDefinition getProgramDefinition(String id);
+  /**
+   * Get the definition for a given program.
+   *
+   * @param id the ID of the program to retrieve
+   * @return an {@link Optional} of the {@link ProgramDefinition} for the given ID if it exists, or
+   *     {@code Optional.empty()} if not
+   */
+  Optional<ProgramDefinition> getProgramDefinition(long id);
 
-    /** Create a new program. */
-    public ProgramDefinition createProgram(String name, String description);
+  /**
+   * Create a new program.
+   *
+   * @param name a name for this program
+   * @param description the description of what the program provides
+   * @return the {@link ProgramDefinition} that was created
+   */
+  ProgramDefinition createProgram(String name, String description);
 
+  /**
+   * Adds a {@link BlockDefinition} to the given program.
+   *
+   * @param programId the ID of the program to update
+   * @param blockName a name for the block to add
+   * @param blockDescription a description of what the questions in the block address
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition addBlockToProgram(long programId, String blockName, String blockDescription);
+
+  /**
+   * Insert a new {@link BlockDefinition} at a specific index for a given program.
+   *
+   * @param programId the ID of the program to update
+   * @param blockIndex the index in the block list for the new block
+   * @param blockName a name for the block to add
+   * @param blockDescription a description of what the questions in the block address
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition addBlockToProgram(
+      long programId, int blockIndex, String blockName, String blockDescription);
```

Comment:
Suggestion (non-blocking): recommend removing this. Specifying block order isn't in-scope for M1 and it's unclear what the semantics/UI we'll be supporting is. Even with a simple action like this there are questions that are tricky to answer like if the index collides with an existing block what happens?

Reply:
Good point - removed